man/procmeterrc.5
.\"  ProcMeter - A system monitoring program for Linux - Version 3.6.
.\"
.\"  Manual page for .procmeterrc file
.\"
.\"  Written by Andrew M. Bishop
.\"
.\"  This file Copyright 1998-2012 Andrew M. Bishop
.\"  It may be distributed under the GNU Public License, version 2, or
.\"  any higher version.  See section COPYING of the GNU Public license
.\"  for conditions under which this file may be redistributed.
.\"
.TH PROCMETERRC 5 "December 30, 2011"

.SH NAME

procmetetrc \- The procmeterrc configuration file for procmeter3,
procmeter3-xaw, gprocmeter3, procmeter3-gtk1, procmeter3-gtk2, procmeter3-gtk3,
procmeter3-log and procmeter3-lcd programs.

.SH SYNOPSIS

This file contains the run-time options for the procmeter3, procmeter3-xaw,
gprocmeter3, procmeter3-gtk1, procmeter3-gtk2, procmeter3-gtk3, procmeter3-log
and procmeter3-lcd programs.  These include the outputs that are displayed, the
colours and fonts that are used (for X window outputs), the LCDproc server (for
the procmeter-lcd version) and other program options.

.SH DESCRIPTION

The resources for ProcMeter version 3.x are stored in the .procmeterrc file
instead of using X resources.  This gives more flexibility and allows for much
simpler customisation.
.LP
When the procmeter3, procmeter3-xaw, gprocmeter3, procmeter3-gtk1,
procmeter3-gtk2, procmeter3-gtk3, procmeter3-log or procmeter3-lcd programs
start the
.B .procmeterrc
file is searched for in the current directory, the current users home directory
or the system default installation directory (named procmeterrc).  The file is
used from the first one of these locations that matches.
.LP
The format of the
.B .procmeterrc
file is a number of sections with the section names marked in '[]'.  As well as
the fixed sections (described below) there is a section for each module that is
named using the name that the module reports itself to be.  There are then
further sections for each of the possible outputs of each module.  The names of
the sections and parameters are not case sensitive although the values that they
are given may be.  Lines beginning with the characters ';' or '#' are comments
and continuation using the '\\' character at the end of a line is allowed.
.LP
Below is listed the resource values and the sections of the
.B .procmeterrc
file that they can exist in.

.SH library SECTION

This section contains the information about the ProcMeter modules that are
available.  This applies to all versions of the program.
.TP
.BR path=...
The pathname of the library that contains the compiled ProcMeter modules.
(defaults to the program's compiled-in installation location).
.TP
.BR others=...
A list of other modules to be loaded (full pathnames separated by spaces).
.LP

.SH startup SECTION

This section controls how the ProcMeter window should look when the program is
started.
.TP
.BR order=...
A list of the outputs that are to be started up in the order that they are to be
displayed (separated by spaces).  If command line options are also specified
then they appear after those specified here.

.SH resources SECTION

This section contains the general program display resources.  Not all of the
options apply to all of the versions of the program.  Specific versions that the
options apply to are specified below.
.TP
.BR horizontal=...
A Boolean to indicate that the outputs are stacked horizontally instead of
the default which is vertically. (For X versions of ProcMeter only.)
.TP
.BR background=...
The colour to use as the default background for all outputs. (An X11 colour name
or an RGB value in hex as #RRGGBB; for X versions of ProcMeter only.)
.TP
.BR foreground=...
The colour to use as the default foreground for all outputs. (An X11 colour name
or an RGB value in hex as #RRGGBB; for X versions of ProcMeter only.)
.TP
.BR label-font=...
The font to use for the label that describes the output. (An X11 font name, also
known as an X Logical Font Description; for Xaw and GTK1/2 versions of
ProcMeter only.)
.TP
.BR label-font-pango=...
The font to use for the label that describes the output. (A Pango font name; for
GTK3 version of ProcMeter only.)
.TP
.BR label-foreground=...
The colour to use as the foreground for the label. (An X11 colour name or an RGB
value in hex as #RRGGBB; for X versions of ProcMeter only.)
.TP
.BR label-position=...
The position of the label relative to the body of the output.  (Either top,
bottom or none; for X versions of ProcMeter only.)
.TP
.BR grid-foreground=...
The colour to use for the grid lines. (An X11 colour name or an RGB value in hex
as #RRGGBB; for X versions of ProcMeter only.)
.TP
.BR graph-solid=...
A boolean to indicate if the graph should be solid (the default) or a single
line.  (yes or no, true or false, 0 or 1; for X versions of ProcMeter only.)
.TP
.BR grid-min=...
The minimum number of grid lines to display on a graph or bar chart.  (An
integer; for X and LCD versions of ProcMeter only.)  This is particularly
important with the bar chart because with this set too low the bar will always
be the same length on screen and the number of grid lines will change with the
data.
.TP
.BR grid-max=...
The maximum number of grid lines to display on a graph or bar chart.  (An
integer; for X and LCD versions of ProcMeter only.)
.TP
.BR text-font=...
The font to use for the textual outputs.  (An X11 font name, also known as an X
Logical Font Description; for Xaw and GTK1/2 versions of ProcMeter only.)
.TP
.BR text-font-pango=...
The font to use for the textual outputs.  (A Pango font name; for GTK3 version
of ProcMeter only.)
.TP
.BR menu-foreground=...
The colour to use as the foreground for the menus.  (An X11 colour name or an RGB
value in hex as #RRGGBB; for Xaw version of ProcMeter only.)
.TP
.BR menu-background=...
The colour to use as the background for the menus.  (An X11 colour name or an
RGB value in hex as #RRGGBB; for Xaw and GTK3 versions of ProcMeter only; the
background of all items on the same menu must be the same.)
.TP
.BR menu-font=...
The font to use for the menus.  (An X11 font name, also known as an X
Logical Font Description; for Xaw version of ProcMeter only.)
.TP
.BR update=...
The update interval for the output in seconds (an integer).
.LP

.SH LCD SECTION

This section contains parameters that are only used by the procmeter3-lcd
version of the program.
.TP
.BR host=...
The name of the host that the LCDd daemon is running on.
.TP
.BR port=...
The port number that the LCDd daemon is running on.
.TP
.BR duration=...
The length of time that all of the ProcMeter outputs are displayed on the LCD
display for before cycling to the next.
.TP
.BR priority=...
The priority of all of the ProcMeter outputs when they are displayed on the LCD
display.  This values should be a numeric value for LCDd verions before 0.5 or
priority names for later versions.  Using numeric values will likely fail for
LCDd versions after 0.5 due to a bug in LCDd.  The safest solution is not to use
this option unless you are sure what value to use.

.SH MODULE SECTION

The name of the section is the name of the ProcMeter module (see the help,
procmeter3 \-h), for example:
.I ProcMeter
.LP
These can contain any of the resource values which will then apply to all
outputs of this module.  There are also extra options that can be applied to the
module and as defaults for the all of the outputs of that module.
.TP
.BR options=...
A string containing options that are passed to the module at startup, different
modules have different options, see the procmeter3_modules(1) manual page.
.TP
.BR graph-scale=...
The scaling of the graph or bar chart, the number of units between each of the
grid lines.  (An integer; for X versions of ProcMeter only.)
.TP
.BR run=...
A string that specifies a program that can be run by selecting it from the menu
for the left mouse button.  (For X versions of ProcMeter only.)  The way that
the command is run can be specified using one of the following abbreviations;
.B XTerm(
.I command
.B )
to run
.I command
in an
.B xterm
then exit.
.B XTermWait(
.I command
.B )
to run
.I command
in an
.B xterm
and wait for user input before exiting.
.B Shell(
.I command
.B )
to run
.I command
in an
.B shell
and exit.  This last one is equivalent to using none of these abbreviations.
.LP

.SH OUTPUT SECTION

The name of the section is the name of the ProcMeter output (see the help,
procmeter3 \-h), for example:
.I ProcMeter.Version
.LP
The output sections can contain any of the resource or module options (except
\'options') which will then apply only to the specified output.  There are also
options that can only be applied to individual outputs.
.TP
.BR label=...
Changes the default label for the output to the specified string.

.SH EXAMPLE

Below is an example
.B .procmeterrc
file, it is a much reduced version of the system default procmeterrc file.

 [startup]

 # ProcMeter version, clock, CPU graph, load text
 order=ProcMeter.Version Date_Time.Time_HM \\
       Statistics.CPU-g Processes.Load-t

 [resources]

 # The outputs, black on white
 foreground = black
 background = white

 # A large-medium size text font.
 text-font = 8x13
 text-font-pango = Sans 8

 # Solid graphs with at least 5 grid lines in grey.
 graph-solid = yes
 grid-min = 5
 grid-foreground = grey50

 # Black on white output labels, small font, below the data.
 label-font = 5x7
 label-font-pango = Sans 5
 label-foreground = black
 label-position = bottom

 # Black on white menu items in a small-medium size font.
 menu-foreground = black
 menu-background = white
 menu-font = 7x13

 [ProcMeter.Version]

 # A smaller font than normal with no label.
 text-font = 6x10
 text-font-gtk3 = Sans 6
 label-position = none

 [Statistics]

 [Statistics.CPU]

 # The minimum number of grid lines, they are 20% each.
 grid-min = 5

 [Processes]

 run = XTerm(top)

 [Network]

 # To pick up extra devices not automatically recognised.
 #options=ppp0 slip0

.SH FILES

.B ./.procmeterrc
.LP
.B ~/.procmeterrc
.LP
.B /usr/local/lib/X11/ProcMeter3/procmeterrc
or
.B /usr/lib/X11/ProcMeter3/procmeterrc

.SH SEE ALSO

procmeter3(1), procmeter3-xaw(1), gprocmeter3(1), procmeter3-gtk1(1),
procmeter3-gtk2(1), procmeter3-gtk3(1), procmeter3-log(1), procmeter3-lcd(1),
procmeter3_modules(1).

.SH AUTHOR

Andrew M. Bishop (c) 1998-2011


NEWS
Version 3.5d of ProcMeter3 released : Mon Oct 3 18:00:00 2011
-------------------------------------------------------------

Bug Fixes
 Remove gcc-4.4 compilation warnings.
 Make error messages consistent.
 Fix several errors in manual pages.
 Don't crash if /proc/interrupts has very long lines.

New features
 Added an option to limit the number of interrupts listed.
 Allow installation of stripped executables.


Version 3.5c of ProcMeter3 released : Sat Feb 20 14:00:00 2010
--------------------------------------------------------------

Bug Fixes
 Remove 64-bit compilation warnings and fix variable type mismatch.
 Fixed df module to handle disks bigger than 2TB.
 Fixed sensors module to handle new directory format.
 Pass top-level Makefile variables down to sub directories.

New features
 Sort the module menu into alphabetical order.

New modules
 FanSpeed, Temperature & Voltage - Improved hardware sensors (see below).


**** NOTE ****

The new hardware sensor modules require the use of the libsensors library.

The advantage is that it reads the system sensors.conf file (which can contain
customisation of scaling and names - essential for some sensors).

The existing "Sensors" module raises a warning when used and the module will be
removed in a future version of the program.
 
**** NOTE ****


Version 3.5b of ProcMeter3 released : Thu Oct 23 19:30:00 2008
--------------------------------------------------------------

Bug Fixes
 Be more careful looking for modules (check '.' and '..').
 Set the locale to "C" to avoid problems parsing numbers.
 Ensure that graph-min and graph-max options are used properly.
 Extended the line buffer size for parsing /proc/interrupts.
 Remove fixed size line buffers in modules (realloc more space as needed).
 Allow LCD version to specify priority as a string or numeric value.

New Features
 Extended Window Manager Hints can be specified with a -w option.
 Increase the number of interrupts possible in stat-intr.


Version 3.5a of ProcMeter3 released : Sun Dec 16 15:00:00 2007
--------------------------------------------------------------

Bug Fixes
 Stop the GTK2 version crashing when displayed items are removed.
 Make sure that the right mouse button menu works with no outputs displayed.
 Make the GTK1 and GTK2 version windows resize themselves like Xaw version.
 Updates to manual pages to reflect new executable names.
 Bug fixes for ACPI module.


Version 3.5 of ProcMeter3 released : Tue Nov 20 19:30:00 2007
-------------------------------------------------------------

New features
 Added a gtk2 version of ProcMeter3.
 Renamed the executables but added backwards compatibility links.
 Added DESTDIR option to Makefiles.

**** IMPORTANT ****

The Xaw version is now named procmeter3-xaw instead of procmeter3.  A symbolic
link to the original name will still be created and installed.

The GTK1 version is now named procmeter3-gtk1 instead of gprocmeter3.
The GTK2 version is named procmeter3-gtk2.

If only one or the other GTK version is compiled then a link will be created to
it from the name gprocmeter3.

**** IMPORTANT ****


Version 3.4g of ProcMeter3 released : Sun Sep 9 15:30:00 2007
-------------------------------------------------------------

Bug fixes:
 Quicker default update of the date displays (useful after suspending).
 Uptime now accurately shows system running time even after suspend.
 Fix ACPI battery discharge rates.
 Fix for crashes with stat-cpu outputs on multi-CPU machines.
 Change to use longer integer values for netdev outputs.
 Search other places for hardware sensors information.


Version 3.4f of ProcMeter3 released : Fri Feb 16 09:45:00 2007
--------------------------------------------------------------

Bug fixes:
 The biff module re-reads the inbox if the size or timestamp change.
 The df module uses longer strings for reading from /proc/mounts.


Version 3.4e of ProcMeter3 released : Sat Apr 29 16:00:00 2006
--------------------------------------------------------------

Bug fixes:
 Fix some spelling mistakes in the manual pages.
 Change the default scaling for the DiskUsage outputs to 5 grid lines of 20%.
 Fix some gcc-4.x compilation warnings.
 Fix some ACPI module bugs.

New or changed modules:
 Add in the new CPU statistics (iowait, irq, softirq and steal).
 Add in a display of the CPU clock speed.


Version 3.4d of ProcMeter3 released : Sat Apr 30 16:00:00 2005
--------------------------------------------------------------

Bug fixes:
 Updated ACPI support to handle kernel version 2.6.
 Added support for /dev/mapper in disk statistics.
 Change to using 64-bit variables from /proc/stat.


Version 3.4c of ProcMeter3 released : Sun Sep 5 14:30:00 2004
-------------------------------------------------------------

Bug fixes:
 Make grid-max option work for GTK version.
 Handle kernel 2.6.x better for hardware sensors.


Version 3.4b of ProcMeter3 released : Sun Apr 25 09:10:00 2004
--------------------------------------------------------------

New or changed modules:
 The stat-disk module can display individual disk partitions for kernel 2.6.x.
 The modules stat, stat-disk, sensors and meminfo now work with kernel 2.6.x.
 Handle different header line for hostap wireless driver.

Bug fixes:
 Modules that didn't work with kernel 2.6.x now do (see above).
 Improved some module initialisation functions.


**** IMPORTANT ****

For kernel version 2.6.x the disk statistics are in /proc/diskstats and not
/proc/stat.  All configuration references to Statistics.Disk need to be changed
to Stat-Disk.Disk.

Statistics.Disk -> Stat-Disk.Disk

**** IMPORTANT ****


Version 3.4a of ProcMeter3 released : Sat Jun 28 20:00:00 2003
--------------------------------------------------------------

New or changed modules:
 The stat-disk module allows specifying a list of extra devices in config file.
 The biff module does not change the access time of the mailbox.

Bug fixes:
 The biff module does not change the access time of the mailbox file.

procmeterrc.c
/***************************************
  $Header: /home/amb/CVS/procmeter3/procmeterrc.c,v 1.10 2008-05-05 18:45:17 amb Exp $

  ProcMeter - A system monitoring program for Linux - Version 3.5b.

  Handle the .procmeterrc file.
  ******************/ /******************
  Written by Andrew M. Bishop

  This file Copyright 1998-2008 Andrew M. Bishop
  It may be distributed under the GNU Public License, version 2, or
  any higher version.  See section COPYING of the GNU Public license
  for conditions under which this file may be redistributed.
  ***************************************/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include <sys/stat.h>

#include "procmeter.h"
#include "procmeterp.h"

typedef struct _Section *Section;
typedef struct _Parameter *Parameter;

/*+ A section from the file. +*/
struct _Section
{
 char *name;                    /*+ The name of the section. +*/
 Section next;                  /*+ The next section; +*/
 Parameter first;               /*+ The first of the parameters. +*/
};

/*+ A parameter from the file. +*/
struct _Parameter
{
 char *name;                    /*+ The name of the parameter. +*/
 char *value;                   /*+ The value of the parameter. +*/
 Parameter next;                /*+ The next parameter. +*/
};

/*+ The first section. +*/
Section FirstSection=NULL;


/*++++++++++++++++++++++++++++++++++++++
  Load in the configuration file.

  int *argc The number of command line arguments.

  char **argv The command line arguments.
  ++++++++++++++++++++++++++++++++++++++*/

void LoadProcMeterRC(int *argc,char **argv)
{
 char *rcpath=NULL,*home;
 Section *next_section=&FirstSection;
 Parameter *next_parameter=NULL,prev_parameter=NULL;
 int i;

 /* Find the .procmeterrc file. */

 for(i=1;i<*argc;i++)
    if(!strncmp(argv[i],"--rc=",5))
      {
       rcpath=argv[i]+5;
       for((*argc)--;i<*argc;i++)
          argv[i]=argv[i+1];
      }

 if(!rcpath)
   {
    struct stat buf;

    if(!stat(".procmeterrc",&buf))
       rcpath=".procmeterrc";
    else if((home=getenv("HOME")))
      {
       rcpath=(char*)malloc(strlen(home)+16);

       strcpy(rcpath,home);
       strcat(rcpath,"/.procmeterrc");

       if(stat(rcpath,&buf))
         {free(rcpath);rcpath=NULL;}
      }
   }
 if(!rcpath)
    rcpath=RC_PATH "/procmeterrc";

 /* Read the .procmeterrc file. */

 if(rcpath)
   {
    FILE *rc=fopen(rcpath,"r");
    char *line=NULL;
    size_t length=256;
    int continued=0;

    if(!rc)
      {
       fprintf(stderr,"ProcMeter: Cannot open the configuration file %s for reading.\n",rcpath);
       exit(1);
      }

    while(fgets_realloc(&line,&length,rc))
      {
       char *l=line,*r=line+strlen(line)-1;

       while(*l==' ' || *l=='\t')
          l++;
       while(r>l && (*r=='\n' || *r=='\r' || *r==' ' || *r=='\t'))
          *r--=0;

       if(*l==';' || *l=='#' || l>=r)
          continue;

       if(*l=='[' && *r==']')
         {
          l++;*r--=0;
          *next_section=(Section)malloc(sizeof(struct _Section));
          strcpy((*next_section)->name=(char*)malloc(strlen(l)+1),l);
          (*next_section)->next=NULL;
          (*next_section)->first=NULL;
          next_parameter=&(*next_section)->first;
          next_section=&(*next_section)->next;
          continued=0;
         }
       else if(continued)
         {
          continued=0;

          if(l[strlen(l)-1]=='\\')
            {
             continued=1;
             l[strlen(l)-1]=0;
            }

          prev_parameter->value=(char*)realloc((void*)prev_parameter->value,strlen(l)+strlen(prev_parameter->value)+1);
          strcat(prev_parameter->value,l);
         }
       else if(next_parameter)
         {
          char *equal=strchr(l,'=');
          if(equal)
            {
             char *lr=equal-1,*rl=equal+1;
             *equal=0;
             while(*lr==' ' || *lr=='\t')
                *lr--=0;
             while(*rl==' ' || *rl=='\t')
                rl++;
             equal=rl;
            }

          *next_parameter=(Parameter)malloc(sizeof(struct _Parameter));
          strcpy((*next_parameter)->name=(char*)malloc(strlen(l)+1),l);

          continued=0;

          if(equal)
            {
             if(equal[strlen(equal)-1]=='\\')
               {
                continued=1;
                equal[strlen(equal)-1]=0;
               }
             strcpy((*next_parameter)->value=(char*)malloc(strlen(equal)+1),equal);
            }
          else
             (*next_parameter)->value=NULL;

          (*next_parameter)->next=NULL;
          prev_parameter=*next_parameter;
          next_parameter=&(*next_parameter)->next;
         }
      }

    fclose(rc);
   }

 /* Add in extra command line options. */

 for(i=1;i<*argc;i++)
    if(!strncmp(argv[i],"--",2))
      {
       int j;
       char *equal=strchr(argv[i],'='),*dot;

       if(equal)
         {
          char *section=argv[i]+2,*parameter,*value=equal+1;
          Section this_section=FirstSection;

          *equal=0;
          dot=strchr(section,'.');

          if(dot)
            {
             char *dot2=strchr(dot+1,'.');

             if(dot2) /* 2 dots => (module.output).parameter */
               {
                parameter=dot2+1;
                *dot2=0;
               }
             else     /* 1 dot => section.parameter */
               {
                parameter=dot+1;
                *dot=0;
               }

             while(this_section)
               {
                if(!strcasecmp(section,this_section->name))
                  {
                   Parameter this_parameter=this_section->first;

                   while(this_parameter)
                     {
                      if(!strcasecmp(parameter,this_parameter->name))
                        {
                         strcpy(this_parameter->value=(char*)realloc((void*)this_parameter->value,strlen(value)+1),value);
                         break;
                        }
                      this_parameter=this_parameter->next;
                     }

                   if(!this_parameter)
                     {
                      Parameter new_parameter=(Parameter)malloc(sizeof(struct _Parameter));
                      strcpy(new_parameter->name=(char*)malloc(strlen(parameter)+1),parameter);
                      if(value)
                         strcpy(new_parameter->value=(char*)malloc(strlen(value)+1),value);
                      else
                         new_parameter->value=NULL;

                      new_parameter->next=this_section->first;
                      this_section->first=new_parameter;
                     }

                   break;
                  }

                this_section=this_section->next;
               }

             if(!this_section)
               {
                Section new_section=(Section)malloc(sizeof(struct _Section));
                strcpy(new_section->name=(char*)malloc(strlen(section)+1),section);

                new_section->first=(Parameter)malloc(sizeof(struct _Parameter));
                strcpy(new_section->first->name=(char*)malloc(strlen(parameter)+1),parameter);
                if(value)
                   strcpy(new_section->first->value=(char*)malloc(strlen(value)+1),value);
                else
                   new_section->first->value=NULL;
                new_section->first->next=NULL;

                new_section->next=FirstSection;
                FirstSection=new_section;
               }
            }
         }

       for(j=i,i--,(*argc)--;j<*argc;j++)
          argv[j]=argv[j+1];
      }
}


/*++++++++++++++++++++++++++++++++++++++
  Return the results of a section of the configuration file.

  char *GetProcMeterRC Returns the result string or NULL if none.

  char *section The section of the config file.

  char *parameter The parameter in the section of the config file.
  ++++++++++++++++++++++++++++++++++++++*/

char *GetProcMeterRC(char *section,char *parameter)
{
 Section this_section=FirstSection;

 while(this_section)
   {
    if(!strcasecmp(section,this_section->name))
      {
       Parameter this_parameter=this_section->first;

       while(this_parameter)
         {
          if(!strcasecmp(parameter,this_parameter->name))
             return(this_parameter->value);
          this_parameter=this_parameter->next;
         }
      }

    this_section=this_section->next;
   }

 return(NULL);
}


/*++++++++++++++++++++++++++++++++++++++
  Get the resources for an output.

  char *GetProcMeterRC2 Returns the result string or NULL if none.

  char *module The module name (part of section name).

  char *output The output name (part of section name).

  char *parameter The parameter to search for.
  ++++++++++++++++++++++++++++++++++++++*/

char *GetProcMeterRC2(char *module,char *output,char *parameter)
{
 ProcMeterModule m;
 ProcMeterOutput o;
 char section[sizeof(m.name)+sizeof(o.name)+2];

 strcpy(section,module);
 strcat(section,".");
 strcat(section,output);

 return(GetProcMeterRC(section,parameter));
}


/*++++++++++++++++++++++++++++++++++++++
  Free up the memory that is used in this module.
  ++++++++++++++++++++++++++++++++++++++*/

void FreeProcMeterRC(void)
{
 Section this_section=FirstSection;

 while(this_section)
   {
    Section last_section=this_section;
    Parameter this_parameter=this_section->first;

    while(this_parameter)
      {
       Parameter last_parameter=this_parameter;

       if(this_parameter->name)
          free(this_parameter->name);
       if(this_parameter->value)
          free(this_parameter->value);
       this_parameter=this_parameter->next;

       free(last_parameter);
      }

    this_section=this_section->next;

    free(last_section->name);
    free(last_section);
   }
}
